[Feature Request] Adding additional metadata file to pg WAL

rjayakri opened the pull request:
Creating PR for the issue reported in https://github.com/wal-g/wal-g/issues/865.

x4m commented on PostgreSQL.md:
````diff
@@ -127,6 +127,20 @@ To configure the size of one backup bundle (in bytes). Smaller size causes granu
+    }
+}
+```
+If not specified it will fallback to default setting(no wal metadata generation)
````
Please add an explicit value for  "no metadata". This may be important if environment var overrides config var.

rjayakri replied:
Added the explicit value "NOMETADATA" and updated the PostgreSQL.md

x4m commented on internal/bguploader_test.go:
```diff
@@ -176,8 +176,8 @@ func addTestDataFile(t *testing.T, dir string, i int) {
-func testFilename(i int) string {
-	return fmt.Sprintf("B%010d", i)
+func testFilename(i string) string {
+	return fmt.Sprintf("%08d%08d%08s", 1, 1, i)
```
Will the new functionality work OK when wal-push is invoked for not WAL-formatted name?
PostgreSQL archived 00000007.history for example.
Or from time to time users archive their stuff :)

rjayakri replied:
It picks up any file generated in the directory and creates the metadata file. 
For example
0000000200000007000000C7.partial.json
0000000200000007000000C7.partial.lz4
00000003.history.json
00000003.history.lz4
0000000200000007000000C4.00000028.backup.json
0000000200000007000000C4.00000028.backup.lz4

x4m commented on internal/wal_push_handler.go:
```diff
@@ -2,21 +2,51 @@ package internal
 }
 
+const (
+	WalBulkMetadataLevel       = "BULK"
```
Are other our settings capitalized? I'm not sure.

rjayakri replied:
I took the example from "WALG_LOG_LEVEL": "NORMAL/DEVEL". Open for any suggestion in-terms of naming covention.

rjayakri commented:
@x4m Please advice if any more changes required. Thanks.

x4m commented:
@usernamedt can you please look at this PR too?
I don't have any objections, just don't feel confident enough. You were revisiting metadata stuff recently and, probably, you have a more consistent opinion than mine.

usernamedt commented on internal/wal_push_handler.go:
```diff
@@ -50,6 +80,12 @@ func HandleWALPush(uploader *WalUploader, walFilePath string) {
 	bgUploader.Start()
 	err = uploadWALFile(uploader, walFilePath, bgUploader.preventWalOverwrite)
 	tracelog.ErrorLogger.FatalOnError(err)
+	if err == nil && (viper.IsSet(UploadWalMetadata) || viper.GetString(UploadWalMetadata) == WalNoMetadataLevel) {
```
I guess here you check instead for `viper.GetString(UploadWalMetadata) != WalNoMetadataLevel `?

Actually, let's just set the default value for UploadWalMetadata to "NOMETADATA" to avoid viper.IsSet check.
This can be done in config.go:108

usernamedt commented on internal/wal_push_handler_test.go:
```diff
@@ -0,0 +1,47 @@
+	assert.NoError(t, err)
+}
+
+func TestWalMetadataBulkUploader(t *testing.T) {
```
Add test for NOMETADATA setting too?

rjayakri replied:
Added test for NOMETADATA. Since the s3 mock uploader returns any file present, could not build the proper test cases.

rjayakri replied:
Updated the unit tests. Now it provides proper test case for NOMETADATA.

usernamedt commented on internal/wal_push_handler.go:
```diff
@@ -2,21 +2,51 @@ package internal
+var WalMetadataLevels = []string{WalBulkMetadataLevel, WalIndividualMetadataLevel}
+
+type WalMetadataDescription struct {
+	CreateTime     time.Time `json:"created_time"`
```
createTime or created_time?

rjayakri replied:
Changed the parameter name.

usernamedt commented on internal/wal_push_handler.go:
```diff
@@ -2,21 +2,51 @@ package internal
+	WalNoMetadataLevel         = "NOMETADATA"
+)
+
+var WalMetadataLevels = []string{WalBulkMetadataLevel, WalIndividualMetadataLevel}
```
Add WalNoMetadataLevel here?

rjayakri replied:
Added. Thanks

usernamedt commented on internal/wal_push_handler.go:
```diff
@@ -50,6 +80,12 @@ func HandleWALPush(uploader *WalUploader, walFilePath string) {
 	tracelog.ErrorLogger.FatalOnError(err)
+	if err == nil && (viper.IsSet(UploadWalMetadata) || viper.GetString(UploadWalMetadata) == WalNoMetadataLevel) {
+		// Creating consolidated wal metadata only for bulk option
+		if viper.GetString(UploadWalMetadata) == WalBulkMetadataLevel && walFilePath[len(walFilePath)-1:] == "F" {
```
Add a comment regarding "F" logic?

rjayakri replied:
Added comments explaining the process involved.

usernamedt commented on internal/wal_push_handler.go:
```diff
@@ -59,6 +95,36 @@ func HandleWALPush(uploader *WalUploader, walFilePath string) {
 	}
 } //
 
+func walMetadataBulkUploader(uploader *WalUploader, walFilePath string) {
```
rename to uploadBulkMetadata?

rjayakri replied:
Renamed it.

usernamedt commented on internal/wal_push_handler.go:
```diff
@@ -75,9 +141,46 @@ func uploadWALFile(uploader *WalUploader, walFilePath string, preventWalOverwrit
+		return errors.Wrapf(err, "upload: could not stat wal file'%s'\n", walFilePath)
+	}
+	var walMetadata WalMetadataDescription
+	walMetadataS := make(map[string]WalMetadataDescription)
```
Rename walMetadataS to something more meaningful?

rjayakri replied:
Renamed it.

usernamedt commented on internal/wal_push_handler.go:
```diff
@@ -75,9 +141,46 @@ func uploadWALFile(uploader *WalUploader, walFilePath string, preventWalOverwrit
+	if err == nil && (viper.IsSet(UploadWalMetadata) || viper.GetString(UploadWalMetadata) == WalNoMetadataLevel) {
+		err = uploadWALMetadataFile(uploader, walFilePath)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to upload metadata file")
```
Should we abort the entire WAL file upload if we failed to upload metadata?

rjayakri replied:
WAL metadata may be critical for PITR and that was the reason aborting with error. If agreed, I can change it to log the error and continue the process.

usernamedt reviewed (changes requested):
Needs some fixes

usernamedt commented:
Closed PR accidentaly, sorry :)

Have you tested it with `WALG_UPLOAD_CONCURRENCY` > 1? It seems like the current logic with bundle metadata uploading might not work correctly in this case.

rjayakri commented:
> Closed PR accidentaly, sorry :)
> 
> Have you tested it with `WALG_UPLOAD_CONCURRENCY` > 1? It seems like the current logic with metadata uploading might not work correctly in this case because of bgUploader.

Thanks. I tested it for WALG_UPLOAD_DISK_CONCURRENCY but missed the WALG_UPLOAD_CONCURRENCY.
Fixed the code for WALG_UPLOAD_CONCURRENCY > 1 and added test case.

usernamedt commented on internal/wal_push_handler.go:
```diff
@@ -58,17 +58,19 @@ func (err CantOverwriteWalFileError) Error() string {
 		if err != nil {
 			tracelog.ErrorLogger.Printf("unmark wal-g status for %s file failed due following error %+v", walFilePath, err)
 		}
+		if viper.GetString(UploadWalMetadata) == WalBulkMetadataLevel && walFilePath[len(walFilePath)-1:] == "F" {
```
This looks identical to the line 85.

I think that it might be better to move the whole metadata upload logic to the `WalUploader` and upload the metadata files similarly to WAL delta files.

usernamedt replied:
Also, what if the WAL segment ending with "F" will be uploaded faster than the previous segments?

rjayakri replied:
> This looks identical to the line 85.
> 
> Maybe extract the common logic to avoid duplication? Or even move the whole metadata upload logic to the `WalUploader` and upload the metadata files similarly to WAL delta files.

Thanks. Just for my understanding, to move the metadata logic to https://github.com/wal-g/wal-g/blob/master/internal/wal_uploader.go#L45 ?
wal_receiver_handler.go also uses the UploadWalFile@wal_uploader.go, should it have any clash if the metadata logic is moved there.

rjayakri replied:
> Also, what if the WAL segment ending with "F" will be uploaded faster than the previous segments? Will the current bulk uploading logic work?

The logic was added here https://github.com/wal-g/wal-g/pull/899/files#diff-f66d6177f62076d018bb01e9b1e9056c991e8206cdb9d16db9e817faf4356abfR68 for that purpose only.
The archive_command(postgresql.conf) will be called for every wal-log, including the one uploaded by the bguploder. Since the archive_command is called in serial, the "F" file will be called at last (after completing the check for all the files created before). So if any of the previous files were not uploaded, it will be handled by the event called from archive_command. Hope my explanation and understanding is clear. Thanks

usernamedt replied:
> wal_receiver_handler.go also uses the [UploadWalFile@wal_uploader.go](mailto:UploadWalFile@wal_uploader.go), should it have any clash if the metadata logic is moved there.

Nice catch, I think that it might be a good idea to upload the WAL metadata during wal-receive too.

> The archive_command(postgresql.conf) will be called for every wal-log, including the one uploaded by the bguploder. Since the archive_command is called in serial, the "F" file will be called at last (after completing the check for all the files created before). So if any of the previous files were not uploaded, it will be handled by the event called from archive_command. Hope my explanation and understanding is clear. Thanks

When `WALG_UPLOAD_CONCURRENCY` is higher than 1, WAL-G invokes the BgUploader, which looks for WAL files adjacent to the current WAL file and uploads them concurrently to the storage. So, on a single archive_command call there could be multiple WALs being uploaded.

rjayakri replied:
> > wal_receiver_handler.go also uses the [UploadWalFile@wal_uploader.go](mailto:UploadWalFile@wal_uploader.go), should it have any clash if the metadata logic is moved there.
> 
> Nice catch, I think that it might be a good idea to upload the WAL metadata during wal-receive too. So I guess that becomes one more reason to move the metadata uploading to the wal_uploader.go
> 
> > The archive_command(postgresql.conf) will be called for every wal-log, including the one uploaded by the bguploder. Since the archive_command is called in serial, the "F" file will be called at last (after completing the check for all the files created before). So if any of the previous files were not uploaded, it will be handled by the event called from archive_command. Hope my explanation and understanding is clear. Thanks
> 
> When `WALG_UPLOAD_CONCURRENCY` is higher than 1, WAL-G invokes the BgUploader, which looks for WAL files adjacent to the current WAL file and uploads them concurrently to the storage. So, on a single archive_command call there could be multiple WALs being uploaded. So there is no guarantee that the "F" file will be uploaded exactly after the previous WAL files.

Thanks. I will move the logic to wal_uploader.go so that wal-push and wal-receive can use the metadata logic.

rjayakri commented:
Fixed the duplicate code. Happy to make any required changes to move the code to different block.

usernamedt commented:
```
INFO: 2021/03/19 13:25:18.372627 FILE PATH: 00000001000000030000004D.lz4
ERROR: 2021/03/19 13:25:18.372858 upload: could not Upload 'pg_wal/00000001000000030000004D'
: upload: could not stat wal file'/tmp/00000001000000030000004D'
: stat /tmp/00000001000000030000004D: no such file or directory
2021-03-19 13:25:18.379 +05 [9202] LOG:  archive command failed with exit code 1
2021-03-19 13:25:18.379 +05 [9202] DETAIL:  The failed archive command was: /usr/local/bin/wal-g wal-push pg_wal/00000001000000030000004D --config /Users/usernamedt/wal-g.yaml
INFO: 2021/03/19 13:25:20.761373 FILE PATH: 00000001000000030000004F.lz4
ERROR: 2021/03/19 13:25:20.761720 upload: could not Upload 'pg_wal/00000001000000030000004F'
: upload: could not stat wal file'/tmp/00000001000000030000004F'
: stat /tmp/00000001000000030000004F: no such file or directory
2021-03-19 13:25:20.765 +05 [9202] LOG:  archive command failed with exit code 1
2021-03-19 13:25:20.765 +05 [9202] DETAIL:  The failed archive command was: /usr/local/bin/wal-g wal-push pg_wal/00000001000000030000004F --config /Users/usernamedt/wal-g.yaml
```
Currently the archive command is failing, needs fix

usernamedt commented on internal/wal_receive_handler.go:
```diff
@@ -105,6 +107,7 @@ func HandleWALReceive(uploader *WalUploader) {
 			tracelog.ErrorLogger.FatalOnError(err)
 			err = uploader.UploadWalFile(ioextensions.NewNamedReaderImpl(tlh, tlh.Name()))
 			tracelog.ErrorLogger.FatalOnError(err)
+			uploadBulkMetadata(uploader, filepath.Join(getWalFolderPath(), segment.Name()))
```
Why not move this to the wal_uploader.go too?

rjayakri replied:
Moved that logic to wal_uploader.go for the wal-receive command. Thanks

rjayakri commented:
> ```
> INFO: 2021/03/19 13:25:18.372627 FILE PATH: 00000001000000030000004D.lz4
> ERROR: 2021/03/19 13:25:18.372858 upload: could not Upload 'pg_wal/00000001000000030000004D'
> : upload: could not stat wal file'/tmp/00000001000000030000004D'
> : stat /tmp/00000001000000030000004D: no such file or directory
> 2021-03-19 13:25:18.379 +05 [9202] LOG:  archive command failed with exit code 1
> 2021-03-19 13:25:18.379 +05 [9202] DETAIL:  The failed archive command was: /usr/local/bin/wal-g wal-push pg_wal/00000001000000030000004D --config /Users/usernamedt/wal-g.yaml
> INFO: 2021/03/19 13:25:20.761373 FILE PATH: 00000001000000030000004F.lz4
> ERROR: 2021/03/19 13:25:20.761720 upload: could not Upload 'pg_wal/00000001000000030000004F'
> : upload: could not stat wal file'/tmp/00000001000000030000004F'
> : stat /tmp/00000001000000030000004F: no such file or directory
> 2021-03-19 13:25:20.765 +05 [9202] LOG:  archive command failed with exit code 1
> 2021-03-19 13:25:20.765 +05 [9202] DETAIL:  The failed archive command was: /usr/local/bin/wal-g wal-push pg_wal/00000001000000030000004F --config /Users/usernamedt/wal-g.yaml
> ```
> 
> Currently, the archive command is failing, needs a fix

My apologies. I had PGDATA set in the environment and that was the reason it was working. Will fix it soon

rjayakri commented:
Need some clarification/suggestion on the wal_receiver. Currently implemented logic for wal_receive to create metadata based on the file timestamp on pg_wal/<wal_file_name>. But  wal_receive can run from any other system and may not have access to the pg_wal directory to identify the file stats (to create metadatafile). Can I use the system timestamp (at the time of wal-receive ) instead of checking the wa-file timestamp on pg_wal directory.

usernamedt commented:
> Need some clarification/suggestion on the wal_receiver. Currently implemented logic for wal_receive to create metadata based on the file timestamp on pg_wal/<wal_file_name>. But wal_receive can run from any other system and may not have access to the pg_wal directory to identify the file stats (to create metadatafile). Can I use the system timestamp (at the time of wal-receive ) instead of checking the wa-file timestamp on pg_wal directory.

Since replication is synchronous, the network lag won't affect the monotonic increase of CreatedTime but will add some delay. From my point of view, it is OK. @x4m what do you think?

x4m commented:
> > Need some clarification/suggestion on the wal_receiver. Currently implemented logic for wal_receive to create metadata based on the file timestamp on pg_wal/<wal_file_name>. But wal_receive can run from any other system and may not have access to the pg_wal directory to identify the file stats (to create metadatafile). Can I use the system timestamp (at the time of wal-receive ) instead of checking the wa-file timestamp on pg_wal directory.
> 
> Since replication is synchronous, the network lag won't affect the monotonic increase of CreatedTime but will add some delay. From my point of view, using the system timestamp is OK. @x4m what do you think?

Sure, anyway this CreatedTime does not affect how WALs will be applied by PostgreSQL. As long as we delete these files basing judgment on LSN - CeatedTime safely can be a random number.

rjayakri commented:
> > > Need some clarification/suggestion on the wal_receiver. Currently implemented logic for wal_receive to create metadata based on the file timestamp on pg_wal/<wal_file_name>. But wal_receive can run from any other system and may not have access to the pg_wal directory to identify the file stats (to create metadatafile). Can I use the system timestamp (at the time of wal-receive ) instead of checking the wa-file timestamp on pg_wal directory.
> > 
> > 
> > Since replication is synchronous, the network lag won't affect the monotonic increase of CreatedTime but will add some delay. From my point of view, using the system timestamp is OK. @x4m what do you think?
> 
> Sure, anyway this CreatedTime does not affect how WALs will be applied by PostgreSQL. As long as we delete these files basing judgment on LSN - CeatedTime safely can be a random number.

Thanks for the feedback. Will implement the logic for wal-receive

usernamedt commented on internal/configure.go:
```diff
@@ -156,6 +156,21 @@ func getWalFolderPath() string {
+		return dataFolderPath
+	}
+
+	return DefaultDataFolderPath
```
let's use this function in the above getWalFolderPath() too

rjayakri replied:
Thanks. Simplified the function call

usernamedt commented on internal/wal_push_handler.go:
```diff
@@ -50,7 +50,7 @@ func HandleWALPush(uploader *WalUploader, walFilePath string) {
 	err = uploadWALFile(uploader, walFilePath, bgUploader.preventWalOverwrite)
 	tracelog.ErrorLogger.FatalOnError(err)
-	uploadBulkMetadata(uploader, walFilePath)
+	_ = uploadBulkMetadata(uploader, filepath.Join(getRelativeArchiveDataFolderPath(), filepath.Base(walFilePath)))
```
Why ignore the error here?

usernamedt commented on internal/wal_uploader.go:
```diff
@@ -168,12 +208,13 @@ func uploadBulkMetadata(uploader *WalUploader, walFilePath string) {
 		}
 	}
 	dtoBody, _ := json.Marshal(walMetadataArray)
```
Why ignore the error here and above?
```
file, _ := ioutil.ReadFile(walMetadataFile)
walMetadataFiles, _ := filepath.Glob(walMetadataBulkFolder.GetFilePath("")...

```

rjayakri replied:
There was still open discussion about  whether we need to fail if the wal-metadata upload fails for any reason (missing file/upload failure etc). I will change all the function calls based on the decision.

usernamedt commented:
I've slightly refactored the code:
- moved all metadata related logic to the wal_metadata_uploader.go
- removed WALG_RECEIVE_BULKMETADATA_PATH. wal-g already has the logic (in getRelativeArchiveDataFolderPath) to fall back to using /tmp folder when pg_wal/pg_xlog is not available so I think we can utilize it. 
- changed the TestWalPush_HandleWALPush to test if the WAL file has been successfully uploaded to the mock storage

usernamedt commented:
@rjayakri If I understood correctly, you plan to add the WAL metadata info to the `wal-show` command? If so, I think we can merge this PR now and add some integration tests later.
